Add std.experimental.typecons.Final

wilzbach opened the pull request:
This is a revival of #3862 (which was an revival of #2881).
The status from @andralex was:

> For more detail: we should define std.experimental.typecons and put this in it. With that I preapprove this addition. Thanks!

Unfortunately @JakobOvrum seems to be away for the last few months, but we shouldn't let this great PR die ;-)

I copy/pasted the addition to `std.experimental.typecons`, added the missing import. Full credit of course goes to @JakobOvrum (luckily git allows to set the history accordingly).

I also included the changelog entry in a separate commit.

AndrejMitrovic commented on std/experimental/typecons.d:
```diff
@@ -802,3 +802,223 @@ private template findCovariantFunction(alias finfo, Source, Fs...)
+{
+    import std.typecons : Proxy;
+
+    private T HeadConst_value;
```
I think this should begin lowercase like `headConst_value`, capitalization should be limited to static members or manifest constants IIRC.

AndrejMitrovic commented on std/experimental/typecons.d:
```diff
@@ -802,3 +802,223 @@ private template findCovariantFunction(alias finfo, Source, Fs...)
+        this.HeadConst_value = T(args);
+    }
+
+    // Attaching function attributes gives less noisy error messages
```
This actually sounds like something we should fix in the compiler. Is there a bugzilla entry for this?

AndrejMitrovic commented on std/experimental/typecons.d:
```diff
@@ -802,3 +802,223 @@ private template findCovariantFunction(alias finfo, Source, Fs...)
         enum ptrdiff_t findCovariantFunction = x;
 }
+
+/++
```
Nitpick: `/+ +/` comments should be used when needed (e.g. when nesting is required), for ddoc stuff we use `/** */` so we should keep it consistent.

AndrejMitrovic commented on std/experimental/typecons.d:
```diff
@@ -802,3 +802,223 @@ private template findCovariantFunction(alias finfo, Source, Fs...)
+    /// Ditto
+    alias HeadConst_get this;
+
+    auto ref T opUnary(string op)()
```
Missing docs.

AndrejMitrovic commented:
I like it a lot. There were some comments in the previous PR, not sure if they were addressed. E.g. the question about [allowing non-reference types](https://github.com/dlang/phobos/pull/3862#issuecomment-199164913).

wilzbach commented:
> There were some comments in the previous PR, not sure if they were addressed. E.g. the question about allowing non-reference types. 

AFAICT the idea from @andralex was:

> OK. A good step forward is to put this in std.experimental, there's subtleties about it. Then we can defer zeroing in on a name.
> For more detail: we should define std.experimental.typecons and put this in it. With that I preapprove this addition. Thanks!

(that's why I initially didn't make any changes except required, but yeah addressing style nits makes sense -> done)

FYI: std.experimental.typecons already existed.

andralex commented on std/experimental/typecons.d:
```diff
@@ -802,3 +802,224 @@ private template findCovariantFunction(alias finfo, Source, Fs...)
+         + underlying value of type $(D T) except for these mutating operators,
+         + which are disabled.
+         +/
+        @disable void opAssign(Other)(Other other);
```
Can you also hoist `@disable` along with the other attributes?

andralex commented on std/experimental/typecons.d:
```diff
@@ -802,3 +802,224 @@ private template findCovariantFunction(alias finfo, Source, Fs...)
+    alias headConst_get this;
+
+    /// Ditto
+    auto ref T opUnary(string op)()
```
Any unary operator that could reasonably return by reference? Even if a UDT defines it as such, observing `ref` creates a hole in `HeadConst`. Let's go with value return.

andralex commented on std/experimental/typecons.d:
```diff
@@ -802,3 +802,224 @@ private template findCovariantFunction(alias finfo, Source, Fs...)
+    }
+}
+
+/// Ditto
```
For the sake of docs, I suggest massaging this inside the main definition. The downside is that all significant code gets indented one level, but there's not a lot of it.

andralex commented on std/experimental/typecons.d:
```diff
@@ -802,3 +802,224 @@ private template findCovariantFunction(alias finfo, Source, Fs...)
+    assert(a.i == 24);
+}
+
+/// $(D HeadConst) can also be used to create read-only data fields without using transitive immutability:
```
Backquotes for no-whitespace D code?

aG0aep6G commented on changelog.dd:
```diff
@@ -340,6 +342,22 @@ $(LI $(LNAME2 generate, `std.range.generate` fixed to be a proper range.)
 )
 
+$(LI $(LNAME2 headconst, `HeadConst` was added to `std.experimental.typecons`.)
+    $(P $(XREF experimental, typecons, HeadConst) can't be mutated directly.
```
No XREF, please. Use REF instead: `$(REF HeadConst, std,experimental,typecons)`.

andralex commented:
Good modulo nits. Approved, @wilzbach you may merge. I suggest using `Final` and `makeFinal`.

wilzbach commented on changelog.dd:
```diff
@@ -332,14 +334,30 @@ static assert( isFinal!(C.ff));
+    $(P $(REF generate, std, range) was set up to return a different value on each
     call to `front`. In addition, `popFront` did nothing. This means that
-    manipulation functions like $(XREF range, drop) did nothing. The new
+    manipulation functions like $(REF drop, std, range) did nothing. The new
```
@aG0aep6G I copied XREF from the entry above, so I fixed that one too.

wilzbach commented:
>  Attaching function attributes gives less noisy error messages
> This actually sounds like something we should fix in the compiler. Is there a bugzilla entry for this?

(@AndrejMitrovic sorry - your comment disappeared once I addressed your nits.)
Now there is: https://issues.dlang.org/show_bug.cgi?id=16444

> I suggest using Final and makeFinal.

There was a huge debate about this on #3862 and on #2881. Imho important is that this _finally_ finds it way to Phobos and the name shouldn't stop it, especially with the `experimental` buffer before.
-> I can understand your reasoning behind `Final` -> renamed

>  Approved, @wilzbach you may merge. 

Hehe the second time you approve this ;-)
I think it's a bad practice to merge one's own PR & addressing your nits introduced quite a bit of changes, so maybe someone else wants to have the joy of merging this old, but still fancy addition?

andralex commented:
Auto-merge toggled on

andralex commented:
alrighty then
